### DISCUSSION TITLE
Adding simplification rules to the decompiler code

### DISCUSSION THREAD

**Discussion by Wall-AF** [NationalSecurityAgency/ghidra#5433] [Q&A]
I often see expressions in the decompiled code of the form:
`if (a >= b && a != b) { ...`
or
`for (...; a <= b && a != b; ...) { ...`

How can I create and add a rule to the decompiler code that would simplify expressions like this to:
`if (a > b) { ...`
and
`for (...; a < b; ...) { ...`

**Comment by Wall-AF** [comment]
Mmm, seems `RuleLessNotEqual` exists but isn't working in my cases!  Question is WHY!!

**Comment by astrelsky** [comment]
> I often see expressions in the decompiled code of the form:
> `if (a >= b && a != b) { ...`
> or
> `for (...; a <= b && a != b; ...) { ...`
> 
> How can I create and add a rule to the decompiler code that would simplify expressions like this to:
> `if (a > b) { ...`
> and
> `for (...; a < b; ...) { ...`
> 

These definitely look like something that shouldn't occur.

**Comment by LukeSerne** [comment]
> I often see expressions in the decompiled code of the form: [...]

Yeah, the decompiler _should_ be able to simplify that. I'd be curious to know why it doesn't. Do you have an example where this happens? If so, could you share it?

In cases like these, where you want to figure out why a specific rule doesn't apply, it's often really helpful to see the state of the data flow graph at different stages in the decompilation process. Since the simplification rules match on specific structures, they don't apply if that structure isn't found. Finding out what structure is actually there is very useful in pinpointing the issue and fixing it. To help with this, I created [DecompVis](https://github.com/LukeSerne/DecompVis/), which reconstructs the data flow graph as it goes through Ghidra's decompilation.

> Mmm, seems `RuleLessNotEqual` exists but isn't working in my cases! Question is WHY!!

`RuleLessNotEqual` matches on `a <= b && a != b`. In your example, we have `a >= b && a != b`, so it's not exactly the same. Because PCODE doesn't have a greater-than operation, this `>=` operator is created by the C-printer, and final, simplified PCODE will probably be something like `b <= a && a != b`. Still, `RuleLessNotEqual` is clever enough to also accept this pattern, so it should still apply. Without more information (specifically, the bytes of the function that causes this weird decompilation), it is impossible for me to find out why.

You could use the DecompVis program yourself, and/or you can upload the function and then I'll also take a look as to why the decompiler is failing.

> How can I create and add a rule to the decompiler code that would simplify expressions like this to: [...]

The simplification rules of the decompiler are written in C++ and are contained in the `/Ghidra/Features/Decompiler/src/decompile/cpp/` folder in the `ruleaction.cc` and `ruleaction.hh` files. Understanding what these rules do can sometimes be a bit hard, and writing them can get repetitive quickly, causing copy-paste errors or other (subtle) bugs. To help with this, I created [RuleChef](https://github.com/LukeSerne/RuleChef/), which takes a high-level description of a simplification rule, and generates C++ code that implements that rule. While it doesn't support complex rules, simple rules like this can be expressed and generated very easily.

Example definition:
```
LessAndNotEqual: BOOL_AND(INT_LESSEQUAL(a, b), INT_NOTEQUAL(a, b)) => INT_LESS(a, b)
```

<details>
<summary>Example output</summary>
<br>

```c++

class RuleSimplifyLessAndNotEqual : public Rule {
public:
  RuleSimplifyLessAndNotEqual(const string &g) : Rule( g, 0, "simplifylessandnotequal") {}      ///< Constructor
  virtual Rule *clone(const ActionGroupList &grouplist) const {
    if (!grouplist.contains(getGroup())) return (Rule *)0;
    return new RuleSimplifyLessAndNotEqual(getGroup());
  }
  virtual void getOpList(vector<uint4> &oplist) const;
  virtual int4 applyOp(PcodeOp *op,Funcdata &data);
};


/// \class RuleSimplifyLessAndNotEqual
///
/// \brief This rule was automatically generated by RuleChef from the expression:
///
/// BOOL_AND(INT_LESSEQUAL(a, b), INT_NOTEQUAL(a, b)) => INT_LESS(a, b)
///
void RuleSimplifyLessAndNotEqual::getOpList(vector<uint4> &oplist) const
{
  oplist.push_back(CPUI_BOOL_AND);
}


int4 RuleSimplifyLessAndNotEqual::applyOp(PcodeOp *op, Funcdata &data)

{

  // Remaining checks on the match expression
  Varnode* autovar_a;
  Varnode* autovar_b;

  // Checks BOOL_AND(INT_LESSEQUAL(a, b), INT_NOTEQUAL(a, b))
  auto check_and_0 = [&](Varnode* autovar_left_0, Varnode* autovar_right_0) -> int4 {
    if (! autovar_left_0->isWritten()) return 0;
    PcodeOp* temp_pcode_0 = autovar_left_0->getDef();
    if (temp_pcode_0->code() != CPUI_INT_LESSEQUAL) return 0;

    // Checks INT_LESSEQUAL(a, b)
    autovar_a = temp_pcode_0->getIn(0);

    autovar_b = temp_pcode_0->getIn(1);


    if (! autovar_right_0->isWritten()) return 0;
    PcodeOp* temp_pcode_1 = autovar_right_0->getDef();
    if (temp_pcode_1->code() != CPUI_INT_NOTEQUAL) return 0;

    // Checks INT_NOTEQUAL(a, b)
    auto check_notequal_0 = [&](Varnode* autovar_left_1, Varnode* autovar_right_1) -> int4 {
      if (autovar_left_1 != autovar_a) return 0;

      if (autovar_right_1 != autovar_b) return 0;

      return 1;
    };

    Varnode* autovar_0 = temp_pcode_1->getIn(0);
    Varnode* autovar_1 = temp_pcode_1->getIn(1);
    if ((! check_notequal_0(autovar_0, autovar_1)) && (! check_notequal_0(autovar_1, autovar_0)))
      return 0;

    return 1;
  };

  Varnode* autovar_2 = op->getIn(0);
  Varnode* autovar_3 = op->getIn(1);
  if ((! check_and_0(autovar_2, autovar_3)) && (! check_and_0(autovar_3, autovar_2)))
    return 0;

  // matched this PcodeOp - replace this with the simplified structure
  data.opSetOpcode(op, CPUI_INT_LESS);
  data.opSetInput(op, autovar_a, 0);
  data.opSetInput(op, autovar_b, 1);

  return 1;
}

```

</details>

Now you can take the output, paste it into `ruleaction.cc`, add the rule to the `actprop` `ActionPool` in `ActionDatabase::universalAction(Architecture *conf)` (in `coreaction.cc`), recompile the decompiler and install it and your new rule should be working. However, in this specific case, you should first figure out what the root cause of the `RuleLessNotEqual` rule not applying is. The rule I provided will not match if the `RuleLessNotEqual` doesn't match, so it won't help you.

**Reply by Wall-AF** [reply]
I've created [this](https://github.com/NationalSecurityAgency/ghidra/files/11937754/dragon_DgnStateContRecog4_t__DoAction.zip) export from the `Decompile:`Panel's `Debug Function Decompilation` menu of an unfortunately large function (sorry), but it has 3 `for` loops where one has reduced but the others haven't.

**Reply by Wall-AF** [reply]
[Here's](https://github.com/NationalSecurityAgency/ghidra/files/12262033/dragon_SumArrayValsForIdx.zip) another example (created as above), but this time there's 2 `for` loops where the 1st is failing and the 2nd succeeds, both in the same small function!

**Reply by LukeSerne** [reply]
Thanks for providing the smaller sample. This analysis really becomes a lot simpler if the sample is as small as possible. Apologies for the late reply.

---

As for the question why `RuleLessNotEqual` doesn't apply to the guard of the first `for` loop: it just barely doesn't. Based on the graph that DecompVis produces (relevant part of the graph is shown below, the nodes that `RuleLessNotEqual` cares about are highlighted in yellow and brown), you would think that it does. However, because the debugging reporting of the decompiler is not complete (see #4963), the graphs produced by DecompVis are sometimes inaccurate.

![image](https://github.com/NationalSecurityAgency/ghidra/assets/32435769/1b8b7703-19dd-4523-a51c-346000662c7d)

The reason that the rule does not apply is that the second input to `INT_LESSEQUAL` has sequence number `0x112816f0:2a9`, whereas the first input to `INT_NOTEQUAL` has sequence number `0x112816f0:2ad`. If the sequence numbers had been the same, the rule would have applied.

```
0x112816f0:2aa: CF(0x112816f0:2aa) = INT_LESSEQUAL( DI(0x112816ed:521), u0x00007980:2(0x112816f0:2a9) )
0x112816f0:2b0: ZF(0x112816f0:2b0) = INT_NOTEQUAL( u0x00007980:2(0x112816f0:2ad), DI(0x112816ed:521) )
```

Looking at the high p-code around those lines, we see how both of these varnodes are assigned:

```
...
0x112816f0:2a9:	u0x00007980:2(0x112816f0:2a9) = *(ram,u0x00003800(0x112816f0:2a8))
0x112816f0:2aa:	CF(0x112816f0:2aa) = DI(0x112816ed:521) <= u0x00007980:2(0x112816f0:2a9)
0x112816f0:2ad:	u0x00007980:2(0x112816f0:2ad) = *(ram,u0x00003800(0x112816f0:2a8))
...
```

They are both values loaded from ram, from the same address.

---

As for the question why the second `for` loop's guard is optimised, it is not optimised by `RuleLessNotEqual`. Instead, the initial structure of the guard of the relevant `CBRANCH` operation is `BOOL_NEGATE(BOOL_OR(INT_LESS(A, B), INT_EQUAL(A, B)))`, where `A = DX(0x112816bb:503)`, `B = DI(0x112816bb:522)`. This structure is simplified to `INT_LESS(B, A)` in two steps:
```
  BOOL_NEGATE(BOOL_OR(INT_LESS(A, B), INT_EQUAL(A, B)))
= { 164: RuleLessEqual }
  BOOL_NEGATE(INT_LESSEQUAL(A, B))
= { 166: RuleBoolNegate }
  INT_LESS(B, A)
```

The reason these two guards are transformed into slightly different p-code is probably because they use slightly different forms of the `CMP` assembly instruction. The guard that is simplified correctly uses `CMP DX, DI` (comparing two registers), whereas the guard that is not simplified correctly uses `CMP word ptr ES:[BX + 0x88], DI` (comparing a register against a value in memory). The p-code for the latter instruction loads the value from memory several times (once for each of the conditional flags), and the decompiler does not realise they would have the same value. Arguably, it shouldn't make that assumption either - the memory may be volatile, or code executing on another thread might change the memory. However, the translation of this instruction to p-code with multiple reads from memory does seem to go against the manual, which shows that first the second operand is subtracted from the first, and then the flags are set based on that result. So the operands of the `CMP` instruction are only used once, and logically, the load from memory should also occur only once.

---

In conclusion, I think that the translation of the `CMP` instruction from x86 to p-code is inaccurate when one of the operands to the `CMP` instruction is a dereferenced value. The translation should only load this value once and set the condition flags based on that loaded value, instead of loading the value once for every condition flag.

**Reply by Wall-AF** [reply]
How do we fix this?   What changes should I make to the `ia.sinc` file?

**Reply by RootCubed** [reply]
You should be able to introduce a temporary value in the :CMP constructor. For instance, for `:CMP rm16,Reg16`, change it to:
`{ local val = rm16; subflags(val,Reg16); local tmp = val - Reg16; resultflags(tmp); }`.
Repeat the same for the other instances where a pointer is the first argument (`spec_rm{8,16,32,64}` and `rm{8,16,32,64}`).
I tried it out and can see the simplification happening with this sample.

**Reply by Wall-AF** [reply]
@LukeSerne thank you for the detailed explanation, very informative and interesting.
____________________
@RootCubed thanks for the solution.  Do you think there are other cmds that would benefit from adding a temp var in the sleigh?
Sould I also update these:
```
:CMP Reg8,rm8       is vexMode=0 & byte=0x3a; rm8 & Reg8 ...                     { subflags( Reg8,rm8   ); local tmp =  Reg8 -    rm8; resultflags(tmp); } 
:CMP Reg16,rm16     is vexMode=0 & opsize=0 & byte=0x3b; rm16 & Reg16 ...        { subflags(Reg16,rm16  ); local tmp = Reg16 -   rm16; resultflags(tmp); }
:CMP Reg32,Rmr32    is vexMode=0 & opsize=1 & byte=0x3b; Reg32 & mod=3 & Rmr32   { subflags(Reg32,Rmr32 ); local tmp = Reg32 -  Rmr32; resultflags(tmp); }
:CMP Reg32,m32      is vexMode=0 & opsize=1 & byte=0x3b; Reg32 ... & m32         { subflags(Reg32,m32   ); local tmp = Reg32 -    m32; resultflags(tmp); }
@ifdef IA64
:CMP Reg64,rm64     is $(LONGMODE_ON) & vexMode=0 & opsize=2 & byte=0x3b; rm64 & Reg64 ...       { subflags(Reg64,rm64 ); local tmp = Reg64 -   rm64; resultflags(tmp); }
@endif
```

@RootCubed one additional quickie, have you any idea how to fix the annoying `CONCAT22` in the below (from the function in this comment https://github.com/NationalSecurityAgency/ghidra/discussions/5433#discussioncomment-6639783)?
`return CONCAT22((int)(nSum >> 0x10),(int)nSum);`
Both the function return and the local var `nSum` are declared as `ulong` (32-bits as spec'd in `x86-16.cspec`!)

**Reply by Wall-AF** [reply]
> ```
> :CMP Reg8,rm8       is vexMode=0 & byte=0x3a; rm8 & Reg8 ...                     { subflags( Reg8,rm8   ); local tmp =  Reg8 -    rm8; resultflags(tmp); } 
> :CMP Reg16,rm16     is vexMode=0 & opsize=0 & byte=0x3b; rm16 & Reg16 ...        { subflags(Reg16,rm16  ); local tmp = Reg16 -   rm16; resultflags(tmp); }
> :CMP Reg32,Rmr32    is vexMode=0 & opsize=1 & byte=0x3b; Reg32 & mod=3 & Rmr32   { subflags(Reg32,Rmr32 ); local tmp = Reg32 -  Rmr32; resultflags(tmp); }
> :CMP Reg32,m32      is vexMode=0 & opsize=1 & byte=0x3b; Reg32 ... & m32         { subflags(Reg32,m32   ); local tmp = Reg32 -    m32; resultflags(tmp); }
> ```
The above needed fixing too.  Not sure about the one below!
> ```
> @ifdef IA64
> :CMP Reg64,rm64     is $(LONGMODE_ON) & vexMode=0 & opsize=2 & byte=0x3b; rm64 & Reg64 ...       { subflags(Reg64,rm64 ); local tmp = Reg64 -   rm64; resultflags(tmp); }
> @endif
> ```

**Comment by ghidracadabra** [comment]
In general it's a best practice in SLEIGH to create such temporary copies whenever possible.

I can add these fixes and look for other instructions to fix in this file, unless @RootCubed is already working on a PR...

**Reply by RootCubed** [reply]
Yeah you can go ahead, I hadn't started on a PR.  One thing to note is that, if I remember correctly from my testing, I wasn't able to reproduce the issue with any other `cmp` instruction (e.g. `cmp [eax], ecx`), so I believe the root cause might have actually been something related to the `segment` SLEIGH instruction/PcodeOp... Though I suppose it can't hurt to add the temporary copies to the others anyway.

**Reply by Wall-AF** [reply]
I believe its mainly an issue with `x86-16` stuff!

In the same/similar vein, I was trying to determine why a simple 32-bit `ulong` (that is halved) is converted to two 16-bit hi/lo values to be subsequently recombined using a `CONCAT22(..., ...)` in the following example!

Assembly (note the use of the operand size override byte (`66`) to use the 32-bit registers!):
```
   1160:0472 c4 5e f0     016           LES        BX,SS:[BP + local_12]
   1160:0475 66 26 8b     016           MOV        EAX,dword ptr ES:[BX + 0x4]
             47 04
   1160:047a 66 d1 e8     016           SHR        EAX,0x1
   1160:047d 66 50        016           PUSH       EAX
   1160:047f 66 26 8b     01a           MOV        EAX,dword ptr ES:[BX + 0x8]
             47 08
   1160:0484 66 5a        01a           POP        EDX
   1160:0486 66 03 c2     016           ADD        EAX,EDX
   1160:0489 66 33 d2     016           XOR        EDX,EDX
   1160:048c 66 26 f7     016           DIV        dword ptr ES:[BX + 0x4]
             77 04
   1160:0491 c4 5e f4     016           LES        BX,SS:[BP + local_e]
                             al = ((field_8 + (field_4 / 2)) / field_4) % 2^8
                             al = Round(field_8 / field_4 + 0.5) % 2^8
   1160:0494 26 88 47 02  016           MOV        byte ptr ES:[BX + 0x2],AL
```
Decompiled output:
```
         local_e->field1_0x2 =
              (byte)((puVar6->field5_0x8 +
                     CONCAT22((uint)(puVar6->field4_0x4 >> 17),(int)(puVar6->field4_0x4 >> 1))
                     ) / puVar6->field4_0x4);
```
Both fields of `puVar6` are defined as `ulong`!

**Comment by ghidracadabra** [comment]
I encountered some similar issues in x64, e.g.:
```
                   long func(long * a, long * b)
                    <IMPOR   
        long        RAX:8    <RETURN>
        long *      RDI:8    a
        long *      RSI:8    b
     00100000 48 8b 06               MOV       RAX,qword ptr [RSI]
     00100003 48 39 07               CMP       qword ptr [RDI],RAX
     00100006 48 0f 4e 07            CMOVLE    RAX,qword ptr [RDI]
     0010000a c3                     RET
```

But of course there could be issues with the `segment` op as well...

**Comment by ghidracadabra** [comment]
Fixed by f030ecb874bacbccf3c2c4a7882edacd058d4a31 which will be included in the upcoming 10.4 release

**Reply by Wall-AF** [reply]
@ghidracadabra unfortunately not quite!
Having incorporated this fix locally and rebuilt Ghdra, I'm still seeing output like
```
            nElems = g_Obj0x47a_DS_9e86.pbytSizes_0x322[i];
            if ((uint)nElems != g_Obj0x47a_DS_9e86.nElemsIn0x472And0x476_0x456 &&
                g_Obj0x47a_DS_9e86.nElemsIn0x472And0x476_0x456 <= (int)(uint)nElems) {
```
from:
```
   11a8:1497 26 8a 07     MOV        AL,byte ptr ES:[BX]
   11a8:149a b4 00        MOV        AH,0x0
   11a8:149c 8b c8        MOV        CX,AX
   11a8:149e 3b 06 dc a2  CMP        AX,word ptr 
                                                   DS:[g_Obj0x47a_DS_9e86.nElemsIn0x472And0x476_0x456]
   11a8:14a2 7e 12        JLE        LAB_11a8_14b6
```
Maybe to do with sizes of variables, `nElems` is a `byte` and `g_Obj0x47a_DS_9e86.nElemsIn0x472And0x476_0x456` is an `int`!

**Reply by LukeSerne** [reply]
What is the raw pcode generated for those instructions? Could you provide a decompiler debugging xml for this function where you're seeing these unoptimised comparisons?

**Reply by Wall-AF** [reply]
[Here's](https://github.com/NationalSecurityAgency/ghidra/files/12701035/dragon_FUN_11a8_12bc.zip) the functions zipped xml.

Below is the pcode:
```
   11a8:1497 26 8a 07                   MOV        AL,byte ptr ES:[BX]
                                                      $U3800:4 = CALLOTHER "segment", ES, BX
                                                      $U7880:1 = LOAD ram($U3800:4)
                                                      AL = COPY $U7880:1
   11a8:149a b4 00                      MOV        AH,0x0
                                                      AH = COPY 0:1
   11a8:149c 8b c8                      MOV        CX,AX
                                                      CX = COPY AX
   11a8:149e 3b 06 dc a2                CMP        AX,word ptr 
                                                   DS:[g_Obj0x47a_DS_9e86.nElemsIn0x472And0x476_0x4
                                                   56]
                                                      $U3800:4 = CALLOTHER "segment", DS, 0xa2dc:2
                                                      $U7980:2 = LOAD ram($U3800:4)
                                                      $U1c880:2 = COPY $U7980:2
                                                      CF = INT_LESS AX, $U1c880:2
                                                      OF = INT_SBORROW AX, $U1c880:2
                                                      $U1c980:2 = INT_SUB AX, $U1c880:2
                                                      SF = INT_SLESS $U1c980:2, 0:2
                                                      ZF = INT_EQUAL $U1c980:2, 0:2
                                                      $Udd80:2 = INT_AND $U1c980:2, 0xff:2
                                                      $Ude00:1 = POPCOUNT $Udd80:2
                                                      $Ude80:1 = INT_AND $Ude00:1, 1:1
                                                      PF = INT_EQUAL $Ude80:1, 0:1
   11a8:14a2 7e 12                      JLE        LAB_11a8_14b6
                                                      $U8700:1 = INT_NOTEQUAL OF, SF
                                                      $U8800:1 = BOOL_OR ZF, $U8700:1
                                                      CBRANCH *[ram]0x11a814b6:4, $U8800:1
```

**Reply by Wall-AF** [reply]
@LukeSerne does the above help?
